rust/lance-encoding/src/repdef.rs

```diff
@@ -0,0 +1,693 @@
+            return SerializedRepDefs::empty();
+        }
+        let has_nulls = builders.iter().any(|b| b.has_nulls());
+        let total_len = builders.iter().map(|b| b.len.unwrap()).sum();
+        let mut context = SerializerContext::new(total_len, has_nulls);
+        debug_assert!(builders
+            .iter()
+            .all(|b| b.num_layers() == builders[0].num_layers()));
+        for layer_index in (0..builders[0].num_layers()).rev() {
+            let layer =
+                Self::concat_layers(builders.iter().map(|b| &b.repdefs[layer_index]), total_len);
+            match layer {
+                RawRepDef::Validity(def) => {
+                    context.record_validity(&def);
+                }
+                RawRepDef::Offsets(rep) => {
+                    context.record_offsets(&rep);
+                }
+                RawRepDef::NoNull(len) => {
+                    context.record_all_valid(len);
+                }
+            }
+        }
+        context.build()
+    }
+}
+
+/// Starts with serialized repetition and definition levels and unravels
+/// them into validity buffers and offsets buffers
+///
+/// This is used during decoding to create the neccesary arrow structures
+#[derive(Debug)]
+pub struct RepDefUnraveler {
+    rep_levels: Option<LevelBuffer>,
+    def_levels: Option<LevelBuffer>,
+    // Current definition level to compare to.
+    current_def_cmp: u16,
+}
+
+impl RepDefUnraveler {
+    /// Creates a new unraveler from serialized repetition and definition information
+    pub fn new(rep_levels: Option<LevelBuffer>, def_levels: Option<LevelBuffer>) -> Self {
+        Self {
+            rep_levels,
+            def_levels,
+            current_def_cmp: 0,
+        }
+    }
+
+    /// Unravels a layer of offsets from the unraveler into the given offset width
+    ///
+    /// When decoding a list the caller should first unravel the offsets and then
+    /// unravel the validity (this is the opposite order used during encoding)
+    pub fn unravel_offsets<T: ArrowNativeType>(&mut self) -> Result<OffsetBuffer<T>> {
+        let rep_levels = self
+            .rep_levels
+            .as_mut()
+            .expect("Expected repetition level but data didn't contain repetition");
+        let mut offsets: Vec<T> = Vec::with_capacity(rep_levels.len() + 1);
+        let mut curlen: usize = 0;
+        let to_offset = |val: usize| {
+            T::from_usize(val)
+            .ok_or_else(|| Error::invalid_input("A single batch had more than i32::MAX values and so a large container type is required", location!()))
+        };
+        if let Some(def_levels) = &mut self.def_levels {
+            assert!(rep_levels.len() == def_levels.len());
+            // This is a strange access pattern.  We are iterating over the rep/def levels and
+            // at the same time writing the rep/def levels.  This means we need both a mutable
+            // and immutable reference to the rep/def levels.
```

Comment:
I do wonder if we could just handle this as a mut slice and have multiple index variables to achieve the same thing. If we wanted to bypass bounds checking, we could add `unsafe` `get_unchecked` calls. That narrows the unsafe a bit.

Reply:
Got it.  Changed to `get_unchecked` and `get_unchecked_mut`